### PR TITLE
Digest encoding

### PR DESCRIPTION
**Pull request by dtebbs**
In clearmatics/zeth#302, the encoding of binary strings as field elements has changed:

a) ordering of components in the "residual bits" element
b) division of bits between "main" (lowest order) element and "residual bits" (highest order)

In fact, the text and diagrams already align with (b) (i.e. clearmatics/zeth#302 brings the code in-line with the spec, allowing slightly simpler unpacking on chain). This PR covers the changes required for (a), and some small fixes in the relevant section. The packing / unpacking description seems to be fully encapsulated in section 3.3.1.

### CODE REVIEW

File: chapters/chap03-sec03.tex [hunk 1]
```diff
@@ -36,25 +36,25 @@ \section{\zeth~statement after primitive instantiation}\label{instantiation:stat
...
+As we consider SNARKs based on arithmetic circuits defined over a prime field, all variables in the constraint system are interpreted as field elements. Nevertheless, as illustrated in~\cref{zeth-protocol:statement}, part of the statement consists of functions whose co-domains are sets of binary strings (which may be longer than the bit representation of elements of the finite field). While a bit (i.e.~$\{0, 1\}$) is an element of $\FF_p$ ($p$ prime), it is important to minimize the number of gates in the arithmetic circuit (for proof generation efficiency), and to minimize the number of input wires (to improve verification time). This can be done by representing fragments of binary strings as the base 2 decomposition of binary elements, thereby ``packing'' binary strings into multiple elements. Converting binary strings into field elements requires the addition of some arithmetic gates (extending the statement to be proven), but reduces the number of primary inputs (reducing the complexity of the SNARK verification carried out on-chain).
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: `This can be done by representing fragments of binary strings as the base 2 decomposition of binary elements` -> `This can be done by representing fragments of binary strings as the base 2 decomposition of field elements`?

File: chapters/chap03-sec03.tex [hunk 2]
```diff
@@ -36,25 +36,25 @@ \section{\zeth~statement after primitive instantiation}\label{instantiation:stat
...
+The cost of $\groth$ zk-SNARK~\cite{groth2016size} proof verification is linear in the number of primary inputs, since each input acts as a scalar in a costly scalar multiplication of a curve point in $\gset_1$. Hence, while packing slightly increases the prover cost -- by adding constraints to the circuit -- it simplifies the verifier's work.
 
-We detail in this section the method by which we encode (resp.~decode) a set of binary strings (resp.~field elements) into a set of field elements (resp.~binary strings). In the rest of this section, the notion of \emph{packing policy} is used to refer to the set of \emph{packing} and \emph{unpacking} functions.
+In this section we detail the method by which we encode (resp.~decode) a set of binary strings to (resp.~from) sets of field elements. In the rest of this section, the notion of \emph{packing policy} refers to the set of \emph{packing} and \emph{unpacking} functions.
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: `In this section` -> `In this section,`?

File: chapters/chap03-sec03.tex [hunk 3]
```diff
@@ -36,25 +36,25 @@ \section{\zeth~statement after primitive instantiation}\label{instantiation:stat
...
-For each element $x$ in this subset, $\alpha_x = \ceil {\len{x} / \bnFieldBitCap}$, represent the number of field elements to encode all of the element's bits. Let $\beta_x = \floor{\len{x} / \bnFieldBitCap}$ represent the number of field elements which are fully ``filled'' and let $\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ denote the number of remaining bits. For simplicity, as we assume $\zvalueLen < \bnFieldBitCap$, we only define the notation $\gamma_\notev = \zvalueLen$ for the public values $\vin$ and $\vout$.
+The primary inputs that consist of binary strings are: the nullifiers $\nfs{}$, the public values $\vin$ and $\vout$, the signature hash $\hsig$ and the authentication tags $\htags{}$.
+
+For a binary string $x$, let $\alpha_x = \ceil {\len{x} / \bnFieldBitCap}$ be the number of field elements required to completely encode $x$ and let $\beta_x = \floor{\len{x} / \bnFieldBitCap}$ be the number of field elements whose capacity is fully used. Let $\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ be the number of remaining or ``residual'' bits remaining after fully using $\beta_x$ field elements.
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: `$\alpha_x = \ceil {\len{x} / \bnFieldBitCap}$` -> `$\alpha_x = \ceil{\len{x} / \bnFieldBitCap}$`

Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: `$\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ be the number of remaining or ``residual'' bits remaining after fully using $\beta_x$ field elements.` -> `$\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ be the number of remaining (or ``residual'') bits remaining after fully using $\beta_x$ field elements.`?

File: chapters/chap03-sec03.tex [hunk 4]
```diff
@@ -36,25 +36,25 @@ \section{\zeth~statement after primitive instantiation}\label{instantiation:stat
...
 
 \begin{example}\label{instantiation:eg:packing}
-    Let $\FF'$ be a prime field which elements are encoded on $5$ bits.
-%
-    If we take a binary string $A \in \bin^7$, such that: $A = (1111011)$, we get: $\alpha_A = 2,\ \beta_A = 1,\ \gamma_A = 2$.
-    We illustrate the result of packing $A$ to field elements~\cref{instantiation:fig:packingA}
+    Let $\FF'$ be a prime field whose elements have $5$ bits of capacity (that is, field elements can represent numbers between $0$ and at least $2^{5}-1$). Let $A \in \bin^7$ be a binary string of length $7$. Then $\alpha_A = 2$, $\beta_A = 1$, and $\gamma_A = 2$. That is, $A$ can be represented as 2 field elements, or as 1 field element with 2 ``residual'' bits.
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: `between $0$ and at least $2^{5}-1$` -> at most (FF' is a prime field, so we may well want to do arithmetic over e.g. 29 - which binary representation is 11101 - and thus obtain a field with 29 elements)

Not sure this sentence, i.e. `(that is, field elements can represent numbers between $0$ and at least $2^{5}-1$)` is necessary here as we just focus on binary string length for the sake of the example.

File: chapters/chap03-sec03.tex [hunk 5]
```diff
@@ -63,12 +63,13 @@ \subsection{Instantiating the packing functions}\label{instantiation:statement:p
     \caption{Packing of string $A$ (see~\cref{instantiation:eg:packing})}\label{instantiation:fig:packingA}
 \end{figure}
 
-We denote by $\resBitsBLen$ the total number of bits which could not fully fill a field element, i.e. the weighted sum of the $\gamma_x$,
+We now consider strategies to pack all primary inputs that are binary strings. A naive approach is to encode each binary string $x$ as $\alpha_x$ field elements. In general, this results in significant waste (and consequently more field elements than necessary), especially when the number of residual bits is small compared to $\bnFieldBitCap$ (see~\cref{instantiation:fig:packing-multi-strings}). An alternative strategy could be to concatenate all binary strings into a single string $y$ and pack this string into $\alpha_y$ field elements. While this approach minimizes the set of unused bits, it significantly increases the complexity of the unpacking operation.
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: `increases the complexity of the unpacking operation` -> I would briefly expand on that to provide more background to the reader who may not realize/remember that this operation is done on-chain (on the EVM) hence why we'd rather avoid adding complexity. In fact, this operation is annoying is carried out on chain because we need to increase "bit shifts" as we decode the elements (the field capacity/EVM word length mismatch "accumulates" along the decoding), and doing bit-wise arithmetic isn't that cheap on-chain.

File: chapters/chap03-sec03.tex [hunk 6]
```diff
@@ -63,12 +63,13 @@ \subsection{Instantiating the packing functions}\label{instantiation:statement:p
...
 \[
     \resBitsBLen = \gamma_\hsig + 2 \cdot \gamma_\notev + \jsin \cdot (\gamma_{\nf{}} + \gamma_{\htag{}} )
 \]
-
-A simple packing strategy would have been to encode each primary input's field $x$ as $\alpha_x$ field elements. However, this leads to a potentially large waste of space when the PRFs and hash digests are slightly longer than $\bnFieldBitCap$ bits (see~\cref{instantiation:fig:packing-multi-strings}). Another strategy would have been to encode the binary string which results from the concatenation of all binary variables. Doing so, however, would make it expensive to carry out the \emph{unpack} procedure (i.e.~retrieve the original binary strings from the set of field elements, see~\cref{zeth-protocol:process-tx}) on-chain. We thus decided to keep the fully filled $\sum_x \beta_x$ field elements which are ``fully'' filled and aggregate the remaining $\resBitsBLen$ ``residual'' bits in a new variable $\resbits$:
+and the lengths, in field elements, of each of the correponding public inputs are
 \begin{align*}
     \nfFLen &= \floor{\prfNfOutLen / \bnFieldBitCap} \\
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: Can't we use the `\beta` notations here?

File: chapters/chap03-sec03.tex [hunk 7]
```diff
@@ -114,45 +117,46 @@ \subsection{Instantiating the packing functions}\label{instantiation:statement:p
...
+        \caption{Algorithm to pack bits into a field element.}\label{zeth-protocol:fig:packing-alg}
     \end{minipage}%
     \begin{minipage}[t]{.6\textwidth}
         \centering
         \procedure{\packResBits{\nfs{}, \vin, \vout, \hsig, \htags{}}}{%
             out \gets []; r \gets \epsilon; \\
             r \gets \slice{\vin}{\floor{\zvalueLen / \bnFieldBitCap} \cdot \bnFieldBitCap}{}; \\
-            r \gets r \concat \slice{\vout}{\floor{\zvalueLen / \bnFieldBitCap} \cdot \bnFieldBitCap}{}; \\
-            r \gets r \concat \slice{\hsig}{\floor{\crhhsigOutLen / \bnFieldBitCap} \cdot \bnFieldBitCap}{}; \\
+            r \gets \slice{\vout}{\floor{\zvalueLen / \bnFieldBitCap} \cdot \bnFieldBitCap}{} \concat r; \\
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: Here again, can't we use the `\beta` notations?

File: chapters/chap03-sec03.tex [hunk 8]
```diff
@@ -114,45 +117,46 @@ \subsection{Instantiating the packing functions}\label{instantiation:statement:p
...
             \pcreturn out;
         }
-        \caption{Packing algorithm in big endian.}\label{zeth-protocol:fig:packing-alg}
+        \caption{Algorithm to pack bits into a field element.}\label{zeth-protocol:fig:packing-alg}
     \end{minipage}%
     \begin{minipage}[t]{.6\textwidth}
         \centering
         \procedure{\packResBits{\nfs{}, \vin, \vout, \hsig, \htags{}}}{%
             out \gets []; r \gets \epsilon; \\
             r \gets \slice{\vin}{\floor{\zvalueLen / \bnFieldBitCap} \cdot \bnFieldBitCap}{}; \\
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: Above you have:
```
    \widetilde{\hsig} \concat \widetilde{\nfs{}} \concat \widetilde{\htags{}} \concat \vin \concat \vout
```

Which does not seem consistent with this algo. Seems like you need to invert vin and vout here.

File: chapters/chap03-sec03.tex [hunk 9]
```diff
@@ -114,45 +117,46 @@ \subsection{Instantiating the packing functions}\label{instantiation:statement:p
...
+The packed primary inputs are 2277 bits long, corresponding to a small space overhead of $\approx 5\%$ unused bits. Moreover, the 143-bit residual bits can be packed into a single field element. As such, the primary inputs are encoded as 9 field elements.
 Finally, the residual bits are formatted as follows,
 \[
-    \underbrace{padding}_{\text{113 bits}} \concat \underbrace{\vin\vphantom{g}}_{64\ bits} \concat \underbrace{\vout\vphantom{g}}_{64\ bits} \concat \underbrace{\hsig}_{3\ bits} \concat \underbrace{\nf{0}}_{3\ bits} \concat \underbrace{\nf{1}}_{3\ bits} \concat \underbrace{\htag{0}}_{3\ bits} \concat \underbrace{\htag{1}}_{3\ bits}
+    \underbrace{padding}_{\text{113 bits}} \concat \underbrace{\hsig}_{3\ bits} \concat \underbrace{\nf{0}}_{3\ bits} \concat \underbrace{\nf{1}}_{3\ bits} \concat \underbrace{\htag{0}}_{3\ bits} \concat \underbrace{\htag{1}}_{3\ bits} \concat \underbrace{\vin\vphantom{g}}_{64\ bits} \concat \underbrace{\vout\vphantom{g}}_{64\ bits}
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: Hum.. based on the algo above, shouldn't it be: `hsig || nf1 || nf0 || htag1 || htag0...`?

Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Fixed this.  Thanks.

File: chapters/chap03-sec03.tex (hunk 8 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Ah yes, `vout` is pushed first.  Fixed. Thanks.

File: chapters/chap03-sec03.tex (hunk 7 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Switched these to use `\beta_*`.  (I also removed the slice in`vout` and `vin` since they are always fully packed alongside the residual bits.)

File: chapters/chap03-sec03.tex (hunk 6 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Switched to `\beta_*`

File: chapters/chap03-sec03.tex (hunk 5 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: I've expanded this a bit. This extra complexity could be handled as complex logic to compute dynamic shifts and masks (higher gas cost) or as many hard-coded unpack operations (minimal, if any, extra gas cost but more code and less maintainable).  I've tried to capture this.

File: chapters/chap03-sec03.tex (hunk 4 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: So, F_29 would have "capacity" 4 (since it can represent all values $0$ up to $2^4$, but not all values up to $2^5$), and "bit length" 5.  In any case, I've removed this since we define "capacity" elsewhere.

File: chapters/chap03-sec03.tex (hunk 3 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Reworded (I realised I used the word "remaining" later in the sentence, so the first use was redundant).

Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Fixed. Thanks.

File: chapters/chap03-sec03.tex (hunk 2 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Changed.

File: chapters/chap03-sec03.tex (hunk 1 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Changed as suggested.

File: chapters/chap03-sec03.tex [hunk 10]
```diff
@@ -36,25 +36,25 @@ \section{\zeth~statement after primitive instantiation}\label{instantiation:stat
...
-For each element $x$ in this subset, $\alpha_x = \ceil {\len{x} / \bnFieldBitCap}$, represent the number of field elements to encode all of the element's bits. Let $\beta_x = \floor{\len{x} / \bnFieldBitCap}$ represent the number of field elements which are fully ``filled'' and let $\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ denote the number of remaining bits. For simplicity, as we assume $\zvalueLen < \bnFieldBitCap$, we only define the notation $\gamma_\notev = \zvalueLen$ for the public values $\vin$ and $\vout$.
+The primary inputs that consist of binary strings are: the nullifiers $\nfs{}$, the public values $\vin$ and $\vout$, the signature hash $\hsig$ and the authentication tags $\htags{}$.
+
+For a binary string $x$, let $\alpha_x = \ceil{\len{x} / \bnFieldBitCap}$ be the number of field elements required to completely encode $x$ and let $\beta_x = \floor{\len{x} / \bnFieldBitCap}$ be the number of field elements whose capacity is fully used. Let $\gamma_x = \len{x} \pmod{\bnFieldBitCap}$ be the number of remaining or ``residual'' bits remaining after fully using $\beta_x$ field elements.
```
Comment (AntoineRondelet) [inline chapters/chap03-sec03.tex]: Did you address https://github.com/clearmatics/zeth-specifications/pull/3#discussion_r521229171 here?

File: chapters/chap03-sec03.tex (hunk 4 above)
Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: Ah you changed the initial example (in which I didn't talk - on purpose - about "field capacity" as per the implementation).
When I wrote
> "Let $\FF'$ be a prime field which elements are encoded on $5$ bits."

I meant: `\FF' = \FFx{p}` where `p` is prime such that `ceil(log(p,2)) = 5`
Didn't want to add potential confusion with field capacity (i.e. `floor(log(p,2))`) here as it is was not necessary to illustrate the "packing" mechanism.

File: chapters/chap03-sec03.tex (hunk 5 above)
Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: Yes, thanks.

When you say:
> This significantly increases the complexity of the unpacking operation that must be performed on-chain, resulting in a higher gas cost (due to extra logic) or more contract code (if each unpack operation is hard-coded)

Isn't 2 (i.e. `more contract code (if each unpack operation is hard-coded)`) already encapsulated in 1 (i.e. `resulting in a higher gas cost (due to extra logic)`)? Having more code => more OPCODES to pay gas for (and more stack operations) on-chain

File: chapters/chap03-sec03.tex (hunk 6 above)
Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: Thanks

File: chapters/chap03-sec03.tex (hunk 4 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: > I meant: \FF' = \FFx{p} where p is prime such that ceil(log(p,2)) = 5

AFAICT, there is no such $p$ which would also be consistent with $\gamma_A = 2$ and the diagram.
TBH, I was confused that the original does not mention the $floor(log(p,2)) = 5$ requirement, despite giving `\alpha`, `\beta`, `\gamma` etc for A.

We could perhaps give a concrete $p$ with 5 bits capacity if you prefer?  And avoid mentioning capacity?

Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: > AFAICT, there is no such $p$ which would also be consistent with $\gamma_A = 2$ and the diagram.

I probably overlooked the example, but basically I'm saying:
Take, say \FFx{31} (31 = (11111)_2), now you have a string A = 1111011 which cannot fit in 5 bits, so the 5 first (or last depending on the packing function definition) bits become a field element and the 7 mod 5 = 2 last bits become the 2nd field elements.

> there is no such $p$ which would also be consistent with $\gamma_A = 2$ and the diagram.

I think I see what you mean: in the example above (see first half of comment) I compute gamma as `7 mod 5` instead of `7 mod 4` (the capacity). In which case, yes I agree there is an inconistency. I didn't want to rely on the "capacity" to explain how packing worked in the example but you are right, this is necessary.

> We could perhaps give a concrete $p$ with 5 bits capacity if you prefer? And avoid mentioning capacity?

Sure, we can rephrase it to raise any ambiguity :)

File: chapters/chap03-sec03.tex (hunk 5 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: I may be missing something, but it's not clear to me that the gas cost would have to be significantly greater.  I think it's a bit implementation-dependent, hence I origianally just called it "complexity", rather than mentioning on-chain "cost".

Our (current) code does 2 shifts/masks per unpack.  Using the single string approach this is pretty much what would be required per unpack (except in some extreme cases where a string spans 3 field elements).  Unless I'm missing something, the bulk of any extra gas cost would come from dynamically computing *how much* to shift /mask with. So I didn't want to call this a purely "on-chain" cost, since technically these values could be hard-coded.

The second point tries to illustrate that. If the shifts / masks are hard-coded, the *complexity* is then in the source code (which, admittedly, would be pretty terrible) but at runtime the number of shifts / masks performed would not be so different from the current implementation (2 shifts / masks per unpack).  There would be more instructions in the contract (because each unpack would be "unrolled") so deployment might be marginally more expensive, but the number of instructions executed (and thereby the gas cost per tx) is not necessarily greater.

... if that makes sense.  Happy to rephrase, though.  I didn't want to get into too much detail in the text, but didn't want to imply that this would always result in much more expensive transactions.

File: chapters/chap03-sec03.tex (hunk 4 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: > Sure, we can rephrase it to raise any ambiguity :)

I think $p = 41$ is the smallest suitable $p$, so we could replace with something like this?:
```
Consider binary strings $A \in \bin^7$ of length 7, to be encoded over the field $\FFx{41} $.  Field elements ar encoded over 6 bits, and can reliably encode up to $5$ bits.  This gives $\alpha_A = 2$, $\beta_A = 1$, and $\gamma_A = 2$. That is, $A$ can be represented as 2 field elements, or as 1 field element with 2 ``residual'' bits.
```
The above still mentions "Field elements can reliably encode up to $5$ bits" (required in order to give tha alpha, beta, gamma), so may not be any simpler than just saying "capacity".  
TBH, I'm OK either way.

File: chapters/chap03-sec03.tex (hunk 10 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Ah ... I think I lost a change somewhere.  Good catch.

File: chapters/chap03-sec03.tex (hunk 3 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: Should be fixed properly now with `the number of remaining or ``residual'' bits remaining after fully using $\beta_x$ field elements.` -> `the number of ``residual'' bits remaining after fully using $\beta_x$ field elements.`.
Sorry about that.

File: chapters/chap03-sec03.tex (hunk 5 above)
Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: Makes sense, thanks

File: chapters/chap03-sec03.tex (hunk 4 above)
Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: Perfect, let's use:
> Consider binary strings $A \in \bin^7$ of length 7, to be encoded over the field $\FFx{41} $. This field has a capacity of $5$ bits which gives $\alpha_A = 2$, $\beta_A = 1$, and $\gamma_A = 2$. That is, $A$ can be represented as 2 field elements, or as 1 field element with 2 ``residual'' bits.

File: chapters/chap03-sec03.tex (hunk 3 above)
Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: Thanks

File: chapters/chap03-sec03.tex (hunk 4 above)
Comment (dtebbs) [reply chapters/chap03-sec03.tex]: updated this example.

Comment (AntoineRondelet) [reply chapters/chap03-sec03.tex]: Great, thanks @dtebbs !

**Review by AntoineRondelet — APPROVED** [review]
LGTM - thanks @dtebbs

**Comment by AntoineRondelet** [comment]
Also related to #1
